File: app/src/main/java/hedera/hgc/hgcwallet/app_intent/TransferRequestParams.java
package hedera.hgc.hgcwallet.app_intent;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.json.JSONObject;


import hedera.hgc.hgcwallet.modals.HGCAccountID;

public class TransferRequestParams implements IntentParams, UriConvertible, QRConvertible {
    public HGCAccountID account;
    public long amount = 0;
    public String note, name;
    public boolean notify;

    public TransferRequestParams(HGCAccountID accountID) {
        this.account = accountID;
    }

    @Nullable
    public static TransferRequestParams from(@NonNull Uri uri) {
        String action =  uri.getQueryParameter("action");
        if(action != null && action.equals("payRequest")) {
            HGCAccountID account = HGCAccountID.Companion.fromString(uri.getQueryParameter("acc"));
            if (account != null) {
                TransferRequestParams params = new TransferRequestParams(account);
                try {
                    params.amount = Long.parseLong(uri.getQueryParameter("a"));
                } catch (Exception e){
                }
                params.name = uri.getQueryParameter("name");
                params.note = uri.getQueryParameter("n");
                params.notify = uri.getBooleanQueryParameter("nr",false);
                return params;
            }
        }
        return null;
    }

    public static TransferRequestParams from(JSONObject jsonObject) {
        if (jsonObject != null) {
            String action = jsonObject.optString("action","");
            if (action.equals("payRequest")) {
                HGCAccountID accountID = HGCAccountID.Companion.fromString(jsonObject.optString("acc"));
                if (accountID != null) {
                    TransferRequestParams params = new TransferRequestParams(accountID);
                    params.name = jsonObject.optString("name");
                    params.note = jsonObject.optString("n");
                    params.amount = jsonObject.optLong("a");
                    return params;
                }
            }

        }
        return null;
    }

    @Nullable
    public static TransferRequestParams fromBarCode(@NonNull String code) {
        Uri uri = Uri.parse(code);
        if (uri != null) {
            return from(uri);
        }
        return null;
    }


    @NonNull
    @Override
    public String asQRCode() {
        return asUri().toString();
    }

    @NonNull
    @Override
    public Uri asUri() {
        Uri.Builder builder = new Uri.Builder();
        builder.scheme(IntentParams.APP_URL_SCHEMA)
                .authority(IntentParams.APP_HOST)
                .appendPath(IntentParams.APP_URL_PATH)
                .appendQueryParameter("action","payRequest")
                .appendQueryParameter("acc",account.stringRepresentation());
        if (amount > 0) {
            builder.appendQueryParameter("a","" + amount);
        }
        if (name != null && !name.isEmpty()) {
            builder.appendQueryParameter("name",name);
        }
        if (note != null && !note.isEmpty()) {
            builder.appendQueryParameter("n",note);
        }
        if (notify) {
            builder.appendQueryParameter("nr","1");
        }
        return builder.build();
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/app_intent/TransferRequestParams.kt
/*
 *
 *  Copyright 2019 Hedera Hashgraph LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package hedera.hgc.hgcwallet.app_intent

import android.net.Uri

import org.json.JSONObject


import hedera.hgc.hgcwallet.modals.HGCAccountID
import javax.annotation.meta.When

class TransferRequestParams(val account: HGCAccountID) : IntentParams, UriConvertible, QRConvertible {
    var amount: Long = 0
    var note: String? = null
    var name: String? = null
    var notify: Boolean = false


    override fun asQRCode(): String {
        return asUri().toString()
    }

    override fun asUri(): Uri {
        return Uri.Builder().apply {
            scheme(IntentParams.APP_URL_SCHEMA)
            authority(IntentParams.APP_HOST)
            appendPath(IntentParams.APP_URL_PATH)
            appendQueryParameter("action", "payRequest")
            appendQueryParameter("acc", account.stringRepresentation())
            if (amount > 0)
                appendQueryParameter("a", "" + amount)

            if (!name.isNullOrEmpty())
                appendQueryParameter("name", name)

            if (!note.isNullOrEmpty())
                appendQueryParameter("n", note)

            if (notify)
                appendQueryParameter("nr", "1")

        }.build()

    }

    companion object {

        fun from(uri: Uri): TransferRequestParams? {
            return uri.getQueryParameter("action")?.let { action ->
                when (action) {
                    "payRequest" -> {
                        HGCAccountID.fromString(uri.getQueryParameter("acc"))?.let { account ->
                            TransferRequestParams(account).apply {
                                try {
                                    amount = java.lang.Long.parseLong(uri.getQueryParameter("a")!!)
                                } catch (e: Exception) {
                                }

                                name = uri.getQueryParameter("name")
                                note = uri.getQueryParameter("n")
                                notify = uri.getBooleanQueryParameter("nr", false)
                            }
                        }
                    }
                    else -> null
                }
            }
        }

        fun from(jsonObject: JSONObject?): TransferRequestParams? {
            return jsonObject?.let { json ->
                when (json.optString("action", "")) {
                    "payRequest" -> {
                        HGCAccountID.fromString(json.optString("acc"))?.let { accountID ->
                            TransferRequestParams(accountID).apply {
                                name = json.optString("name")
                                note = json.optString("n")
                                amount = json.optLong("a")
                            }
                        }
                    }
                    else -> null
                }
            }
        }

        fun fromBarCode(code: String): TransferRequestParams? {
            return Uri.parse(code)?.let { from(it) }
        }
    }
}
